File: Data Structure I/Day 3/best-time-to-buy-and-sell-stock/solution.cs
// I know this repo is for TS, but I really wanted to make sure I understand this problem

public class Solution {
    public int MaxProfit(int[] prices) {
        var l = prices.Length - 1;
        var profit = 0;
        var currentMin = 100001;
        
        for (int i = 0; i <= l; i++) {
            if (prices[i] < currentMin) {
                currentMin = prices[i];
                continue;
            }
            if (prices[i] - currentMin > profit) {
                profit = prices[i] - currentMin;
            }
        }
        
        return profit;
    }
}

File: Data Structure I/Day 3/best-time-to-buy-and-sell-stock/solution.ts
// O(n^2)

function maxProfit(prices: number[]): number {
    let l = prices.length - 1;
    let max = 0;
    
    for (let i = 0; i < l; i++) {
        for (let j = i+1; j <= l; j++) {
            if (prices[j] - prices[i] > max) {
                max = prices[j] - prices[i];
            }
        }
    }
        
    return max;
};

// O(n)

function maxProfit2(prices: number[]): number {
    let l = prices.length - 1;
    let profit = 0;
    let currentBuy = 100001;
        
    for (let i = 0; i <= l; i++) {
        if (prices[i] < currentBuy) {
            currentBuy = prices[i];
        } else {
            if (prices[i] - currentBuy > profit) {
                profit = prices[i] - currentBuy;
            } 
        }
    }
    
    return profit;
};